hxtabletrans.1
.TH "HXTABLETRANS" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxtabletrans \- transpose an HTML or XHTML table
.SH SYNOPSIS
.B hxtabletrans
.RB "[\| " \-c " \|]"
.RB "[\| " \-x " \|]"
.RB "[\| " \-v " \|]"
.RI "[\| " file-or-URL " \|]"
.SH DESCRIPTION
.B hxtabletrans
reads an HTML or XHTML file (or fragment), finds the first <table>
element and outputs that element with the rows and columns
transposed. I.e., rows become columns and columns become rows.
.PP
Any grouping of rows with <thead>, <tbody> or <tfoot> is lost. Also,
any comments outside the cells are lost.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-c
Also try to transpose the contents of cells. If a cell consists only
of horizontal or vertical arrows and ellipses (apart from mark-up and
whitespace), then those arrows and ellipses are replaced by their
transposed counterparts. E.g., a right arrow becomes a down arrow.
.TP
.B \-x
Use XML syntax. Empty elements are written with a slash (/) at the
end, e.g.: <br/>
.TP
.B \-v
Print the version number and exit.
.SH BUGS
.B hxtabletrans
should check for rows that have fewer cells than the longest row
(taking into account the effect of the colspan and rowspan attributes)
and create the omitted cells, but it currently doesn't. Tables with
rows of unequal length may thus not be transposed correctly.
.PP
Nested tables are not transposed.


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxtabletrans.c
/*
 * hxtabletrans -- transpose an HTML or XHTML table
 *
 * The input is an HTML or XHTML table, the output is the same table
 * transposed, i.e., rows become columns and columns become rows, as
 * if the table was flipped along the main diagonal.
 *
 * Any TBODY, THEAD, COLGROUP are lost. Comments and processing
 * instructions occurring outside of table cells are also lost.
 *
 * TODO: reinsert comments (at some reasonable place).
 *
 * TODO: handle rows that have too few cells. (Requires adding up
 * colspans and rowspans, possibly discounting overlap because of
 * incorrect rowspans.)
 *
 * TODO: Turn thead, tfoot and tbody into colgroup?
 *
 * Copyright © 2012 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/2002/copyright-software-20021231
 *
 * Author: Bert Box <bert@w3.org>
 * Created: 23 Oct 2012
 */
#if HAVE_CONFIG_H
# include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <err.h>
#include <unistd.h>
#include <assert.h>
#include <stdbool.h>
#if HAVE_STRING_H
# include <string.h>
#elif HAVE_STRINGS_H
# include <strings.h>
#endif
#include <ctype.h>
#include "heap.e"
#include "types.e"
#include "tree.e"
#include "scan.e"
#include "html.e"
#include "dict.e"
#include "openurl.e"

static Tree tree;
static bool has_error = false;
static bool do_xml = false;
static bool do_content = false;	/* Whether to try and transpose characters */


/*======================== Parser callbacks =======================*/


/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  warnx("%s:%d: %s\n", get_yyin_name(), lineno, s);
  has_error = true;
}


/* start -- called before the first event is reported */
/* void* handle_start(void) */


/* end -- called after the last event is reported */
/* void handle_end(void *clientdata) */


/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, const string commenttext)
{
  tree = append_comment(tree, commenttext);
}


/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, const string text)
{
  tree = append_text(tree, text);
}


/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, const string gi, const string fpi,
		 const string url)
{
  tree = append_declaration(tree, gi, fpi, url);
}


/* handle_pi -- called after a Processing Instruction is parsed */
void handle_pi(void *clientdata, const string pi_text)
{
  tree = append_procins(tree, pi_text);
}


/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, const string name, pairlist attribs)
{
  tree = html_push(tree, name, attribs);
  free(name);
}


/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, const string name, pairlist attribs)
{
  tree = html_push(tree, name, attribs);
  free(name);
}


/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, const string name)
{
  tree = html_pop(tree, name);
  free(name);
}


/* handle_endincl -- called at the end of an included file */
/* void handle_endincl(void *clientdata) */


/*==================== End of parser callbacks ====================*/


/* print_start -- print a start tag */
static void print_start(Tree t)
{
  pairlist p;

  printf("<%s", t->name);
  for (p = t->attribs; p; p = p->next)
    if (p->value) printf(" %s=\"%s\"", p->name, p->value);
    else printf(" %s", p->name);
  if (do_xml && is_empty(t->name)) printf("/>"); else printf(">");
}


/* print_tree -- print the subtree t */
static void print_tree(Tree t)
{
  if (t) {
    switch (t->tp) {
    case Element:
      print_start(t);
      if (!is_empty(t->name)) {
	print_tree(t->children);
	printf("</%s>", t->name);
      }
      print_tree(t->sister);
      break;
    case Text:
      printf("%s", t->text);
      print_tree(t->sister);
      break;
    case Procins:
      printf("<?%s>", t->text);
      print_tree(t->sister);
      break;
    case Declaration:
      printf("<!DOCTYPE %s", t->name);
      if (t->text && t->url) printf(" PUBLIC \"%s\" \"%s\"", t->text, t->url);
      else if (t->text) printf(" PUBLIC \"%s\"", t->text);
      else if (t->url)  printf(" SYSTEM \"%s\"", t->url);
      printf(">");
      print_tree(t->sister);
      break;
    case Comment:
      printf("<!--%s-->", t->text);
      print_tree(t->sister);
      break;
    case Root:
    default:
      assert(!"Cannot happen!");
      break;
    }
  }
}


/* find_all_rows -- create an array pointing to the 1st child of each row */
static int find_all_rows(Tree t, Tree **rows)
{
  int nrows = 0;
  Tree g, h;

  assert(t->tp == Element && eq(t->name, "table"));
  *rows = NULL;
  for (h = t->children; h; h = h->sister) {
    if (h->tp != Element) continue;
    if (eq(h->name, "tr")) {
      renewarray(*rows, ++nrows);
      (*rows)[nrows-1] = h->children;
    } else if (eq(h->name, "thead") ||
	       eq(h->name, "tbody") ||
	       eq(h->name, "tfoot")) {
      for (g = h->children; g; g = g->sister) {
	if (h->tp != Element) continue;
	assert(eq(g->name, "tr"));
	renewarray(*rows, ++nrows);
	(*rows)[nrows-1] = g->children;
      }
    } else {
      assert(eq(h->name, "colgroup") || eq(h->name, "col"));
    }
  }
  return nrows;
}


/* transpose_char -- if c has a transposed form, return it; otherwise NULL */
static conststring transpose_char(conststring *c)
{
  static conststring pairs[] = {
    "⋮", "⋯",
    "…", "︙",
    "←", "↑",
    "→", "↓",
    "⇐", "⇑",
    "⇒", "⇓",
    "⇠", "⇡",
    "⇢", "⇣",
    "⇦", "⇧",
    "⇨", "⇩",
    "⇄", "⇅",
    "⇆", "⇅",
    "⇇", "⇈",
    "⇉", "⇊",
    NULL,
  };
  int i;

  for (i = 0; pairs[i]; i++)
    if (hasprefix(*c, pairs[i])) {
      *c += strlen(pairs[i]);
      return pairs[i % 2 ? i - 1 : i + 1];
    }
  return NULL;
}


/* appendc -- reallocate and append one character */
static void appendc(string *s, int c)
{
  int i;

  if (!s) {new(s); *s = NULL;}
  i = *s ? strlen(*s) : 0;
  renewarray(*s, i + 2);
  (*s)[i] = c;
  (*s)[i+1] = '\0';
}


/* transpose_content -- try to transpose character content, true if success */
static bool transpose_content(const Tree t, Tree *result)
{
  Tree q = NULL, r;
  conststring h, c;
  string s = NULL;

  *result = NULL;
  if (!t) return true;
  if (t->tp == Element && !transpose_content(t->children, &q)) return false;
  if (!transpose_content(t->sister, &r)) {tree_delete(q); return false;}
  if (t->tp == Text) {
    for (h = t->text; *h;)
      if (isspace(*h)) appendc(&s, *(h++));
	else if ((c = transpose_char(&h))) strapp(&s, c, NULL);
	else {dispose(s); tree_delete(q); tree_delete(r); return false;}
  }
  assert(t->tp != Root);
  new(*result);
  (*result)->tp = t->tp;
  (*result)->parent = t->parent;
  (*result)->children = q;
  (*result)->sister = r;
  if (t->tp == Comment || t->tp == Procins || t->tp == Declaration)
    (*result)->text = newstring(t->text);
  else
    (*result)->text = s;
  if (t->tp == Declaration) (*result)->url = newstring(t->url);
  if (t->tp == Element || t->tp == Declaration)
    (*result)->name = newstring(t->name);
  if (t->tp == Element) (*result)->attribs = pairlist_copy(t->attribs);
  return true;
}


/* trans -- print the transposition of table t */
static void trans(const Tree t)
{
  Tree *rows;
  int nrows, i;
  pairlist p;

  static Node _SPANNED;
  static Tree SPANNED = &_SPANNED;
  Tree **table, h;
  int j, k, l, colspan, rowspan, maxcols, *ncols;
  conststring s;
  Tree transposed;

  assert(t->tp == Element && eq(t->name, "table"));
  nrows = find_all_rows(t, &rows);

  /* Find the maximum number of columns */
  /* TODO: deal with overlapping cells */
  newarray(ncols, nrows);
  for (i = 0; i < nrows; i++) ncols[i] = 0;

  for (i = 0; i < nrows; i++) {
    for (h = rows[i]; h; h = h->sister) {
      assert(h->tp == Element || h->tp == Comment || h->tp == Procins);
      if (h->tp == Element) {
	assert(eq(h->name, "td") || eq(h->name, "th"));
	if (! (s = get_attrib(h, "colspan"))) colspan = 1;
	else if ((colspan = atoi(s)) < 1) colspan = 1;
	if (! (s = get_attrib(h, "rowspan"))) rowspan = 1;
	else if ((rowspan = atoi(s)) < 1) rowspan = 1;
	for (j = i; j < i + rowspan && j < nrows; j++) ncols[j] += colspan;
      }
    }
  }
  for (maxcols = 0, i = 0; i < nrows; i++)
    if (maxcols < ncols[i]) maxcols = ncols[i];

  /* Make an empty table */
  newarray(table, nrows);
  for (i = 0; i < nrows; i++) {
    newarray(table[i], maxcols);
    for (j = 0; j < maxcols; j++) table[i][j] = NULL;
  }

  /* Put the cells at their correct coordinates in the table */
  for (i = 0; i < nrows; i++) {
    for (j = 0, h = rows[i]; h; h = h->sister) {
      if (h->tp == Element) {
	while (table[i][j]) j++; /* Find next free cell in this row */
	assert(j < maxcols);
	if (! (s = get_attrib(h, "colspan"))) colspan = 1;
	else if ((colspan = atoi(s)) < 1) colspan = 1;
	if (! (s = get_attrib(h, "rowspan"))) rowspan = 1;
	else if ((rowspan = atoi(s)) < 1) rowspan = 1;
	for (k = i; k < i + rowspan; k++)
	  for (l = j; l < j + colspan; l++)
	    table[k][l] = SPANNED; /* Mark cells as occupied */
	table[i][j] = h;
	j += colspan;
      }
    }
  }

  /* Print the table transposed */
  print_start(t);
  printf("\n");

  for (j = 0; j < maxcols; j++) {
    printf("<tr>\n");
    for (i = 0; i < nrows; i++) {
      if (table[i][j] == NULL) {
	printf("<td></td>\n");
      } else if (table[i][j] != SPANNED) {
	/* Swap colspan and rowspan attributes */
	for (p = table[i][j]->attribs; p; p = p->next)
	  if (eq(p->name, "colspan")) strcpy(p->name, "rowspan");
	  else if (eq(p->name, "rowspan")) strcpy(p->name, "colspan");
	print_start(table[i][j]);
	if (do_content && transpose_content(table[i][j]->children, &transposed))
	  print_tree(transposed);
	else
	  print_tree(table[i][j]->children);
	printf("</%s>\n", table[i][j]->name);
      }
    }
    printf("</tr>\n");
  }
  printf("</table>\n");
}


/* transpose -- look for the first table in the tree and output it transposed */
static bool transpose(const Tree t)
{
  if (!t) return false;
  else if (t->tp == Element && eq(t->name, "table")) {trans(t); return true;}
  else if (transpose(t->sister)) return true;
  else return transpose(t->children);
}


/* usage -- print usage message and exit */
static void usage(const string prog)
{
  fprintf(stderr, "Usage: %s [-c] [-x] [-v] [file-or-url]\n", prog);
  exit(1);
}


/* main -- main body */
int main(int argc, char *argv[])
{
  int c, status = 200;

  /* Parse command line */
  while ((c = getopt(argc, argv, "cxv")) != -1)
    switch (c) {
    case 'c': do_content = true; break;
    case 'x': do_xml = true; break;
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    default: usage(argv[0]);
    }

  if (optind == argc)
    set_yyin(stdin, "stdin");
  else if (optind == argc - 1)
    set_yyin(fopenurl(argv[optind], "r", &status), argv[optind]);
  else
    usage(argv[0]);

  if (!yyin) err(1, "%s", argv[argc-1]);
  if (status != 200) errx(1, "%s: %s", argv[argc-1], http_strerror(status));

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse input and transpose it as a table, if possible */
  tree = create();
  if (yyparse() != 0) exit(1);
  if (!transpose(get_root(tree))) errx(1, "Found no table to transpose");

  return has_error ? 1 : 0;
}
